plugins/http/src/commands.rs

```diff
@@ -20,6 +20,64 @@ pub struct FetchResponse {
     url: String,
 }
 
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct ProxyConfig {
+    all: Option<String>,
+    http: Option<String>,
+    https: Option<String>,
+    basic_auth: Option<BasicAuth>,
+    no_proxy: Option<String>,
+}
+
+#[derive(Deserialize)]
+pub struct BasicAuth {
+    username: String,
+    password: String,
+}
+
+fn attach_proxy(
+    settings: ProxyConfig,
+    builder: reqwest::ClientBuilder,
+) -> crate::Result<reqwest::ClientBuilder> {
+    let ProxyConfig {
+        all,
+        http,
+        https,
+        basic_auth,
+        no_proxy,
+    } = settings;
+
+    let mut builder = builder;
+
+    let attach_auth_and_exclusion = |proxy: Proxy| {
+        let mut attatched = proxy;
+        if let Some(BasicAuth { username, password }) = &basic_auth {
+            attatched = attatched.basic_auth(username, password);
+        }
+        if let Some(no_proxy) = &no_proxy {
+            attatched = attatched.no_proxy(NoProxy::from_string(no_proxy));
+        }
+        attatched
+    };
+
+    if let Some(all_proxy) = all {
+        let proxy = attach_auth_and_exclusion(Proxy::all(all_proxy)?);
+        builder = builder.proxy(proxy);
+    }
+    if let Some(http_proxy) = http {
+        let proxy = attach_auth_and_exclusion(Proxy::http(http_proxy)?);
+        builder = builder.proxy(proxy);
+    }
+    if let Some(https_proxy) = https {
+        let proxy = attach_auth_and_exclusion(Proxy::https(https_proxy)?);
+        builder = builder.proxy(proxy);
+    }
```

Comment:
I think we can remove the need for a closure here by doing something similar to this:
```rs
let proxy = if let Some(all) = all {
	Some(Proxy::all(all)?)
  } else if let Some(https) = https {
    Some(Proxy::https(https)?)
  } else if let Some(http) = http {
    Some(Proxy::http(http)?)
  } else {
    None
  } ;

if let Some(proxy) = proxy {
  builder = builder.proxy(proxy);
  if let Some(BasicAuth { username, password }) = &basic_auth {
    proxy = proxy.basic_auth(username, password);
  }
  if let Some(no_proxy) = &no_proxy {
    proxy= proxy.no_proxy(NoProxy::from_string(no_proxy));
  }
}
```

Reply:
In my personal opinion, wouldn't this implementation only allow for a single proxy to exist

Reply:
ah my bad, I thought you could only add one proxy.